src/handlers/register.rs

```diff
@@ -32,24 +34,89 @@ pub async fn handler(
     claims.verify_basic(&state.auth_aud, None)?;
     let client_id = ClientId::from(claims.iss);
 
+    increment_counter!(state.metrics, register);
+
+    if let Some(tags) = body.tags {
+        increment_counter!(state.metrics, registration_overwrite);
+        overwrite_registration(&state, client_id.clone(), tags, body.relay_url).await?;
+    } else {
+        increment_counter!(state.metrics, registration_update);
+        update_registration(
+            &state,
+            client_id.clone(),
+            body.append_tags,
+            body.remove_tags,
+            body.relay_url,
+        )
+        .await?;
+    }
+
+    Ok(Response::default())
+}
+
+async fn overwrite_registration(
+    state: &Arc<AppState>,
+    client_id: ClientId,
+    tags: Vec<Arc<str>>,
+    relay_url: Arc<str>,
+) -> error::Result<Response> {
     state
         .registration_store
         .upsert_registration(
             client_id.value(),
-            body.tags.iter().map(AsRef::as_ref).collect(),
-            body.relay_url.as_ref(),
+            tags.iter().map(AsRef::as_ref).collect(),
+            relay_url.as_ref(),
         )
         .await?;
 
     state
         .registration_cache
         .insert(client_id.into_value(), CachedRegistration {
-            tags: body.tags.clone(),
-            relay_url: body.relay_url.clone(),
+            tags,
+            relay_url,
         })
         .await;
 
-    increment_counter!(state.metrics, register);
-
     Ok(Response::default())
 }
+
+async fn update_registration(
+    state: &Arc<AppState>,
+    client_id: ClientId,
+    append_tags: Option<Vec<Arc<str>>>,
+    remove_tags: Option<Vec<Arc<str>>>,
+    relay_url: Arc<str>,
+) -> error::Result<Response> {
+    if let (Some(append_tags), Some(remove_tags)) = (append_tags.clone(), remove_tags.clone()) {
+        for tag in remove_tags.iter() {
+            if append_tags.contains(tag) {
+                return Err(Error::InvalidUpdateRequest);
+            }
+        }
+    }
+
+    let registration = state
+        .registration_store
+        .get_registration(client_id.as_ref())
+        .await?;
+
+    let mut tags = registration.tags;
+
+    if let Some(remove_tags) = remove_tags {
+        for r_tag in remove_tags.iter() {
+            if let Some(index) = tags.iter().position(|tag| tag.eq(r_tag)) {
+                tags.remove(index);
+            }
```

Comment:
Will this not result in a "concurrent modification exception" if using Java? I.e. removing while at the same time iterating can result in wrong indexes